kube/src/api/subresource.rs

```diff
@@ -208,3 +210,134 @@ where
+    /// Allocate a terminal for this attach call; defaults to `false`.
+    pub tty: bool,
+}
+
+impl Resource {
+    /// Attach to a pod
+    pub fn attach<R, W>(&self, name: &str, ap: &AttachParams<R, W>) -> Result<http::Request<()>>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite + Unpin,
+    {
+        let base_url = self.make_url() + "/" + name + "/" + "attach?";
+        let mut qp = url::form_urlencoded::Serializer::new(base_url);
+
+        if ap.stdin.is_some() {
+            qp.append_pair("stdin", "true");
+        }
+        if ap.stdout.is_some() {
+            qp.append_pair("stdout", "true");
+        }
+        if ap.stderr.is_some() {
+            qp.append_pair("stderr", "true");
+        }
+        if ap.tty {
+            qp.append_pair("tty", "true");
+        }
+        if let Some(container) = &ap.container {
+            qp.append_pair("container", &container);
+        }
+
+        let req = http::Request::post(qp.finish());
+        req.body(()).map_err(Error::HttpError)
+    }
+}
+
+/// Marker trait for objects that has attach
+pub trait AttachingObject {}
+
+impl AttachingObject for k8s_openapi::api::core::v1::Pod {}
+
+impl<K> Api<K>
+where
+    K: Clone + DeserializeOwned + AttachingObject,
+{
+    /// Attach to pod
+    pub async fn attach<R, W>(&self, name: &str, mut ap: AttachParams<R, W>) -> Result<()>
+    where
+        R: AsyncRead + Unpin,
+        W: AsyncWrite + Unpin,
+    {
+        let req = self.resource.attach(name, &ap)?;
+        let stream = self.client.connect(req).await?;
+
+        let (mut _send, mut recv) = stream.split();
+        let mut p_msg = recv.next();
+
+        loop {
+            // TODO Handle stdin and tty
+            match p_msg.await {
+                Some(Ok(msg)) => {
+                    match msg {
+                        Message::Binary(bin) if !bin.is_empty() => {
+                            // Write to appropriate channel
+                            match bin[0] {
+                                // stdin
+                                0 => {}
+                                // stdout
+                                1 => {
+                                    if let Some(stdout) = ap.stdout.as_mut() {
+                                        stdout.write_all(&bin[1..]).await?;
+                                    }
+                                }
+                                // stderr
+                                2 => {
+                                    if let Some(stderr) = ap.stderr.as_mut() {
+                                        stderr.write_all(&bin[1..]).await?;
+                                    }
+                                }
+                                // error?
+                                3 => {}
+                                // resize?
+                                4 => {}
+                                _ => {}
+                            }
+                        }
```

Comment:
are there any good sources for this protocol?

Reply:
Not really :( I haven't found it.

I'm using the official Python client as a reference for this because I think it has the most complete WebSocket support. The channel numbers are from these lines: https://github.com/kubernetes-client/python-base/blob/fdc9907ff0d8afbc952ad22ec65549b5c4fde0b7/stream/ws_client.py#L33-L37

They're not using `RESIZE_CHANNEL` (4), but TypeScript client has some code that might be related to it https://github.com/kubernetes-client/javascript/blob/5f0c70f03c7de72c59ade79e20dc94726b7a7930/src/attach.ts#L47-L51

---

In general, each message (bytes) is prefixed with a channel like this:

```text
| channel | data |
|   0     | 1..  |
```

For attach/exec, channels are:
```text
0 = stdin, 1 = stdout, 2 = stderr, 3 = error, 4 = resize
```

For `port-forward`, it uses 2 channels per port. Duplex data channel (`2*i`) and error channel (`2*i + 1`) for `i`th port.

There's an initial channel frame that's 3 bytes.

1. channel (u8)
2. port number (u16 BE)

After this, each message has channel prefix like the `attach` one, but the channel is associated with the port.

I haven't looked in to `port-forward` much yet.